### ISSUE TITLE
error in SQLFORM.grid with table auth_membership 

### ISSUE THREAD

**Issue by geordanis** [web2py/web2py#1569]
Error snapshot : <type 'exceptions.TypeError'>(reduce() of empty sequence with no initial value)
The table auth_membership has some data correctly inserted.

**Comment by abastardi** [comment]
I cannot reproduce. Please show your exact code.

**Comment by geordanis** [comment]
Before, thanks for your attention.
the error starts when I trying to use de SQLFOM.grid with the table auth_membership.

I have got this in a controller:
@auth.requires_membership("Administradores")
def admin_panel():
    query = request.vars.query or "auth_user" #to recive the name of the table to be showed, other wise auth_user by default
    grid = SQLFORM.grid(db[query],user_signature=False,editable=True,deletable=True,create=True) 
    return locals()

query=auth_membership:
If I write directly the user name (ej. tom.bush) in the search’s input and the error if showed, if I use the search combobox  to prepare the query (ej. auth_membership.user_id = "4", the id of tom.bush record) a correct output I get.

regards

**Comment by geordanis** [comment]
In my case the built-in restrictions of the table auth_membership respect to the tables auth_user and auth_group  does not work correctly with the SQLFORM.grid search mechanism.

**Comment by abastardi** [comment]
If you just enter a search string directly in the search input without using the advanced search query builder, the `SQLFORM.build_query` method will only search `string`, `text`, and `list:string` fields. However, because the `auth_membership` table does not include any of those field types, `build_query` produces this error. This is a bug in `build_query`, which should check to make sure there is at least one searchable field type in the table before proceeding.

Note, even without that bug, you would not get any search results in this case. The `auth_membership` table includes only reference fields, which store integers. If you enter a user's name as the search keyword, there will be no match, as the names are not stored in the `auth_membership` table. That is why the advanced search widget is there -- it includes dropdowns for the reference fields showing the representations for each foreign key in the field.

**Comment by geordanis** [comment]
That was the answer i was expecting for; thanks a lot. That not records were showed with a search string directly in the input in these cases could happen, but should not happens that error be showed without treatment; i an thinking as a final user not as programmer; so i hope that this bug be removed. regards

**Comment by abastardi** [comment]
I agree -- an exception should not be generated simply because of the content a user enters in the search box.

**Comment by geordanis** [comment]
in sqlhtml.py line 1823 if the table has not searchables fields as the case of auth_membership a error is raised because of
a reduce() with an empty collection

        if key and not '"' in key:
            SEARCHABLE_TYPES = ('string', 'text', 'list:string')
            sfields = [field for field in fields if field.type in SEARCHABLE_TYPES]
            if settings.global_settings.web2py_runtime_gae:
                return reduce(lambda a,b: a|b, [field.contains(key) for field in sfields])
            else:# elif sfields: to avoid an empty list to be passed to the reduce()
                return reduce(lambda a,b:a&b,[
                        reduce(lambda a,b: a|b, [
                                field.contains(k) for field in sfields]
                               ) for k in key.split()])

this way there is not exception, but all records are returned and i think that is not correct
